### PR TITLE
Handle Llama 3.2 chat template

### PR DESCRIPTION
**Pull request by DePasqualeOrg**
This is a work in progress, and I'm not sure how to proceed, since I'm not very familiar with Jinja or this package. I think I managed to solve the test for if `tools` is defined, but now it's failing when it tries to iterate through an undefined `tools`.

### CODE REVIEW

**Comment by maiqingqiang** [comment]
The chat template of Llama3.2 uses tojson, but Swift Jinja has not been implemented yet. You can refer to the commits of jinja.js for porting.

https://github.com/huggingface/huggingface.js/commit/78e6a5eb81c63446c03cba8bd2325d8124c171be
https://github.com/huggingface/huggingface.js/commit/717d69d8ad11e90dde79ebda416d0127ee8dc56a

**Comment by maiqingqiang** [comment]
Would you be willing to become a Swift Jinja Collaborator? I hope to find like-minded partners to co-build Swift Jinja, as my abilities are limited on my own.

**Comment by DePasqualeOrg** [comment]
Sure, I'm happy to help out, although Jinja and the way it's used in these chat templates are still quite new to me.

**Comment by DePasqualeOrg** [comment]
> The chat template of Llama3.2 uses tojson, but Swift Jinja has not been implemented yet. You can refer to the commits of jinja.js for porting.
> 
> [huggingface/huggingface.js@78e6a5e](https://github.com/huggingface/huggingface.js/commit/78e6a5eb81c63446c03cba8bd2325d8124c171be) [huggingface/huggingface.js@717d69d](https://github.com/huggingface/huggingface.js/commit/717d69d8ad11e90dde79ebda416d0127ee8dc56a)

I think this line is not being evaluated as false:

```jinja
{%- if tools is not none %}
```

Because then it tries to iterate over `tools`, which is not iterable when it is `none`.

**Comment by maiqingqiang** [comment]
> > The chat template of Llama3.2 uses tojson, but Swift Jinja has not been implemented yet. You can refer to the commits of jinja.js for porting.
> > [huggingface/huggingface.js@78e6a5e](https://github.com/huggingface/huggingface.js/commit/78e6a5eb81c63446c03cba8bd2325d8124c171be) [huggingface/huggingface.js@717d69d](https://github.com/huggingface/huggingface.js/commit/717d69d8ad11e90dde79ebda416d0127ee8dc56a)
> 
> I think this line is not being evaluated as false:
> 
> ```
> {%- if tools is not none %}
> ```
> 
> Because then it tries to iterate over `tools`, which is not iterable when it is `none`.

Yes, `none` also needs to be implemented.

You can port the commit.
https://github.com/huggingface/huggingface.js/commit/80cd45661ced1df0c1a19d5ee20e96b850c4021d

**Comment by DePasqualeOrg** [comment]
Thank you. Is there a point in time at which this package fully mirrored the JavaScript implementation? That would be helpful to know so that I can see which new additions might need to be added here.

**Comment by maiqingqiang** [comment]
> Thank you. Is there a point in time at which this package fully mirrored the JavaScript implementation? That would be helpful to know so that I can see which new additions might need to be added here.

This package started development on May 20 https://github.com/huggingface/swift-transformers/pull/77#issuecomment-2008571237 , but at that time it did not fully mirror the JavaScript implementation. It only completed the mainstream LLM chat template [Tests/ChatTemplateTests.swift](https://github.com/maiqingqiang/Jinja/blob/main/Tests/ChatTemplateTests.swift) .

 It should be at https://github.com/huggingface/huggingface.js/commit/80cd45661ced1df0c1a19d5ee20e96b850c4021d.

**Comment by pcuenca** [comment]
Both `none` and the use of tools were introduced very recently in jinja.js, to support Llama 3.1. I think that implementing them would go a long way towards supporting most of the chat templates currently in use.

**Comment by DePasqualeOrg** [comment]
These Jinja packages are complex enough that at the moment the best I can do is hand the problem off to Claude 3.5 Sonnet, and that seems to only get part of the way to a solution. So feel free to take it from here if you want to. It would probably take me several days to get up to speed, and I unfortunately don't have enough capacity to do that right now.

**Comment by maiqingqiang** [comment]
It's okay, I can help you solve the problem.

**Comment by DePasqualeOrg** [comment]
Okay, I got this working with Llama 3.2. I haven't implemented the tools part with `tojson`, but it will work for inference in a normal chat scenario. I also haven't added tests, and I can see that one test is already failing.

**Comment by DePasqualeOrg** [comment]
I added a test for `none`, and I had to make some changes from the TypeScript version to make it pass. Please check if this is correct.

**Comment by DePasqualeOrg** [comment]
I checked https://github.com/huggingface/huggingface.js/pull/846 again and fixed a few things that I think now bring this in line with the TypeScript implementation. I think this is now ready for review.

File: Sources/Parser.swift [hunk 1]
```diff
@@ -415,6 +415,9 @@ func parse(tokens: [Token]) throws -> Program {
             current += 1
 
             return ObjectLiteral(value: values)
+        case .nullLiteral:
+            current += 1
+            return NullLiteral()
```
[the remark this answers is not in the document]
Comment (DePasqualeOrg) [reply Sources/Parser.swift]: Yes, thank you. I've removed it in the latest commit.

**Comment by maiqingqiang** [comment]
Thanks a lot for the PR!

**Pull request merged**